README.md
# Ruby in Fantasy

Simple toolbox library and lean API to build great mini games in Ruby.

An upper layer over Gosu library to offer a more friendly API for an easy and lean game development.

Specially intended to use Ruby as a learning language to introduce children into programming.

**Attention**: This project is in early development phase, right now it is just an experiment

## Use

### Hello World

```ruby
require "fantasy"

SCREEN_WIDTH = 320
SCREEN_HEIGHT = 240

on_game do
  message = HudText.new(position: Coordinates.new(10, 10))
  message.text = "Hello, World!"
  message.size = "big"
end

start!
```

### Action game

```ruby
SCREEN_WIDTH = 640
SCREEN_HEIGHT = 360

on_game do
  player = Actor.new("warrior")
  player.position = Coordinates.new(100, 100)
  points = HudText.new(position: Coordinates.new(10, 20))
  points.text = 0

  on_space_bar do
    Sound.play("shoot")
    bullet = Actor.new("bullet")
    bullet.position = player.position
    bullet.speed = 100
    bullet.direction = Coordinates.up
    bullet.on_collision do |other|
      if other.name == "enemy"
        Sound.play("impact")
        other.destroy
        bullet.destroy
        points.text += 1
      end
    end
  end

  player.on_collision do |other|
    if other.name == "enemy"
      Sound.play("game_over")
      Global.go_to_end
    end
  end
end

on_end do
  HudText.new(position: Coordinates.new(10, 100), text: "You are dead. Press space to re-tart")

  on_space_bar do
    Global.go_to_game
  end
end

start!
```

## Examples

See the [Ruby in Fantasy Games Collection](https://github.com/fguillen/RubyInFantasyGames).

## Installation

Add this line to your application's Gemfile:

```ruby
gem "fantasy"
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install fantasy

## Features

### Game Scenes

Easy to configure 3 basic game states:

- Game presentation scene
- Game game scene
- Game end scene
- Other scenes, like levels or such (TODO)

Each state should be independent and unique Actors and other elements can be created and configured for each state

Built-in mechanism to move from one state to another.

### Actor

Managing game elements which have (optionally) image, movement and collision

- Easy to set an `image`
- Managing movement through `direction` and `speed`
- Built-in movement control through cursors
- Collision detection, OnCollision callback, Collision matrix
- Jump
- Gravity
- Animations (TODO)
- Possibility to extend Actor class or instantiate it directly for simple characters
- Allowing magic instance properties (Like in OpenStruct). So programmer can do `actor.stuff = 1` and it is valid (TODO)
- Alignment "center" and "top-left" (TODO)

### Clock

Create new execution threads to program delayed behaviors, repetitive actions, animations, ...

- Run now
- Run after a delay
- Repeat
- Stop
- Callback when finished (TODO)

### Animations

Add animations to an Actor and activate them when special state triggered

- Set frames (TODO)
- Duration in seconds (TODO)
- Run once (TODO)
- Run on loop (TODO)
- Stop (TODO)
- Callback when finished (TODO)

### UI

#### Button (TODO)

Easy to set up a graphical component that responds to mouse hover and click events.

- Set image (TODO)
- Set image when hover (TODO)
- Set image when clicked (TODO)
- OnClicked callback (TODO)

#### HUD Elements

For easy creation of head-up display components.

- Text Element
- Image Element
- Auto update text (observer variable changes) (TODO)

### Debug Mode

Press `d` to activate it. When active debug visuals are shown:

- Position
- Collider

On top, when debug mode is active, the Actors and HUD elements become draggable.
The new position won't be persistent but it will be helpful to identify which should be the desired position.

### Camera

Setting up a camera component that we can easily move around and all the active
Actors in the game will be rendered in the relative position to this camera.

- Allow camera move
- Easy to follow one actor

### Colors palette

- Not deal with RGB or anything, just a list of colors. See [Color.palette](COLOR_PALETTE.md)


### Pause Game (TODO)

- Pause game, pause Actors, animations, clocks, ... (TODO)

### Core run

Move the core functions to the top level hierarchy so I don't need to create a `Game::Window` class

### Sound

- Direct and easy way to play a sound
- Direct and easy way to play a background music


### Background

Simple way to set up:

- Background color
- Image background
- Repeatable image background

### Data Persistance

Simple mechanism to save data in disk. For user preferences, game progress, high scores and others

### User Inputs

Easy access to keyboard and mouse inputs on any part of the code. Specially in the Actors.

- Allow "on_space_bar" and each Actor (TODO)
- Allow multiple "on_space_bar" (TODO)
- Remove "on_space_bar" when changing scene (TODO)
- Detect when key/mouse button is pressed in the actual frame in any part of the code (TODO)

### Tilemap (TODO)

For easy creation of:

- Top-down map levels
- Platformer map levels (TODO)
- Allow specific map keys like a=>actor_1, b=>actor_2 (TODO)

### Tweens (TODO)

Multiple movement animation effects like in [DoTween](http://dotween.demigiant.com/documentation.php) (TODO)

## API

### Game Scenes

Configure your game elements on each Scene:

```ruby
SCREEN_WIDTH = 640
SCREEN_HEIGHT = 360

# (Optional)
on_presentation do
  # Game elements running when the game loads
end

on_game do
  # Game elements running when in game Scene
end

# (Optional)
on_end do
  # Game elements running when game is ended
end
```

How to go from Scene to Scene:

```ruby
Global.go_to_presentation
Global.go_to_game
Global.go_to_end
```

#### Example

```ruby
on_presentation do
  HudText.new(position: Coordinates.new(10, 100), text: "Press space to start")

  on_space_bar do
    Global.go_to_game
  end
end

on_game do
  # [...]
  if player.dead
    Global.go_to_end
  end
end

on_end do
  HudText.new(position: Coordinates.new(10, 100), text: "You are dead. Press space to re-tart")

  on_space_bar do
    Global.go_to_presentation
  end
end
```

### Actor

Actor can be used directly:

```ruby
player = Actor.new("warrior") # ./images/warrior.png
player.position = Coordinates.new(100, 100)
player.solid = true
player.speed = 200
player.layer = 1
player.move_with_cursors
player.collision_with = ["enemy", "bullets"] # default "all"

player.on_collision do |other|
  if other.name == "enemy"
    player.destroy
  end
end

player.on_after_move do
  if player.position.x > Global.screen_width
    player.position.x = Global.screen_width
  end

  if player.position.x < 0
    player.position.x = 0
  end
end
```

Or in a subclass:

```ruby
class Player < Actor
  def initialize
    super("warrior") # ./images/warrior.png
    @position = Coordinates.new(100, 100)
    @solid = true
    @speed = 200
    @layer = 1
    @direction = Coordinates.zero
    @collision_with = ["enemy", "bullets"] # default "all"
    move_with_cursors
  end

  def on_collision_do(other)
    if other.name == "enemy"
      destroy
    end
  end

  def on_after_move_do
    if @position.x > SCREEN_WIDTH
      @position.x = SCREEN_WIDTH
    end

    if @position.x < 0
      @position.x = 0
    end
  end
end
```

Actor settings for a platformer:

```ruby
player = Actor.new("warrior") # ./images/warrior.png
player.solid = true
player.jump = 150
player.gravity = 200
player.move_with_cursors(left: true, right: true, up: false, down: false, jump: true)

player.on_jumping do
  Sound.play("jump")
  player.image = "warrior_jump")
end

player.on_floor do
  player.image = "warrior")
end
```

### Clock

```ruby
clock =
  Clock.new do
    enemy.attack
    sleep(1)
    enemy.defend
  end

clock.run_now
clock.run_on(seconds: 2)
clock.repeat(seconds: 2, times: 10)
clock.stop
```

### Background

```ruby
# Simple color
on_presentation do
  Global.background = Color.new(r: 34, g: 35, b: 35)
end

# Replicable (by default) Image
# position is relative to Camera.main
on_game do
  background = Background.new(image_name: "beach")
  # background.replicable = false # if you don't want the image to replicate
  background.scale = 6
end
```

### Data Persistance

```ruby
Disk.data.records = [120_000, 11_000, 678]
Disk.data.last_level = 3
Disk.data.sound_volume = 12
Disk.save # data stored in ./disk/data.json

# ... in another session

Disk.data.records # => [120_000, 11_000, 678]
Disk.data.last_level # => 3
Disk.data.sound_volume # => 12
```

### Camera

```ruby
on_game do
  on_loop do
    # Camera follows player
    Camera.main.position.y = player.position.y - (SCREEN_HEIGHT / 2)
  end
end
```

### Sound

Plays only once:

```ruby
Sound.play("shoot") # ./sounds/shoot.wav
```

Plays in loop:

```ruby
Music.play("music") # ./musics/music.mp3
Music.volume = 0.5
Music.stop
```

### Tilemap

```ruby
# ./maps/sky.txt
# 0     0
#  0  1
#   0   0
# 0    01

planet = Actor.new("planet")
star = Actor.new("star")
map = Tilemap.new(map_name: "sky", tiles: [planet, star], tile_size: 30)
map.spawn
```

### UI

#### HUD Text

```ruby
timer = HudText.new(position: Coordinates.new(SCREEN_WIDTH / 2, 10))
timer.text = 0
timer.alignment = "center"
timer.size = "big"

Clock.new { timer.text += 1 }.repeat(seconds: 1)
```

#### HUD Image

```ruby
icon = HudImage.new(position: Coordinates.new(SCREEN_WIDTH - 220, 8), image_name: "ring")
icon.scale = 4
icon.visible = true

Clock.new { icon.visible = !icon.visible }.repeat(seconds: 1)
```


## Credits for assets

- Font: VT323 Project Authors (peter.hull@oikoi.com)

## Bugs

- When dragging in debug mode new elements are being added to the drag (TODO)
- Rubocop is not passing
- Tests are missing
- Allow Ruby 2.5+

## Assets

- Images for the test fixtures: https://kenney.nl/

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/fguillen/fantasy.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
## [0.1.13] - 2022-04-01

- Data Persistance: Disk module
- Sound and Music support for volume configuration


## [0.1.11] - 2022-03-26

- Better jumping and gravity physics
- Support Shape.rectangle
- Support on_click callback on Shapes and Actors
- Support on_cursor_* callbacks on Shapes and Actors
- Support Color.palette with 150 colors + transparent


## [0.1.9] - 2022-03-25

- Adding platformer support
  - Jump
  - Gravity
- Music background support


## [0.1.7] - 2022-03-17

- HudText accepts alignment property. Valid values: 'top-left, top-right, center'


## [0.1.5] - 2022-03-16

- Background color or image
- Tilemap
- Camera can be controlled by cursors
- Background image can be replicable


## [0.1.0] - 2022-03-11

- Initial release


## [Unreleased]


lib/fantasy/actor.rb
# frozen_string_literal: true

# Represents one active entity in the game.
# An actor can be the Player sprite. Or one of the enemies. Or the bullet.
# An actor can also be a tree, or a wall
#
# @example Actor can be used directly
#   player = Actor.new("warrior") # ./images/warrior.png
#   player.position = Coordinates.new(100, 100)
#   player.solid = true
#   player.speed = 200
#   player.layer = 1
#   player.move_with_cursors
#   player.collision_with = ["enemy", "bullets"] # default "all"
#
#   player.on_collision do |other|
#     if other.name == "enemy"
#       player.destroy
#     end
#   end
#
#   player.on_after_move do
#     if player.position.x > Global.screen_width
#       player.position.x = Global.screen_width
#     end
#
#     if player.position.x < 0
#       player.position.x = 0
#     end
#   end
#
# @example Or in a subclass:
#   class Player < Actor
#     def initialize
#       super("warrior") # ./images/warrior.png
#       @position = Coordinates.new(100, 100)
#       @solid = true
#       @speed = 200
#       @layer = 1
#       @direction = Coordinates.zero
#       @collision_with = ["enemy", "bullets"] # default "all"
#       move_with_cursors
#     end
#
#     on_collision do |other|
#       if other.name == "enemy"
#         destroy
#       end
#     end
#
#     on_after_move do
#       if @position.x > Global.screen_width
#         @position.x = Global.screen_width
#       end
#
#       if @position.x < 0
#         @position.x = 0
#       end
#     end
#   end
class Actor
  include MoveByCursor
  include MoveByDirection
  include Mover
  include Gravitier
  include Jumper
  include UserInputs

  # Coordinates object where x and y represent the position of the Actor in the World (no necessarily in the Screen).
  #
  # Default `Coordinates.zero`.
  #
  # @return [Coordinates] the actual position
  #
  # @example Setting position
  #   actor = Actor.new("image")
  #   actor.position = Coordinates.new(10, 20)
  #   actor.position.x # => 10
  #
  # @example Modify position
  #   actor.position.x +=  1
  #   actor.position.x # => 11
  attr_accessor :position

  # Controls the direction in which this Actor will move in the next frame.
  #
  # Default `Coordinates.zero`.
  #
  # @return [Coordinates] the actual direction
  #
  # @note The the pixels per second is represented by the `@speed` attribute
  #
  # @example Set direction
  #   actor = Actor.new("image")
  #   actor.direction = Coordinates.right
  attr_accessor :direction

  # Controls the pixels per second which this Actor will move in the next frame.
  #
  # Default `0`.
  #
  # @return [Float] the actual speed
  #
  # @note The the direction is represented by the `@direction` attribute
  #
  # @example Set speed
  #   actor = Actor.new("image")
  #   actor.speed = 10
  attr_accessor :speed

  # Controls impulse this Actor will receive when the (see #jump) is triggered.
  #
  # Default `0`.
  #
  # @return [Float] the actual jump_force
  #
  # @note The the direction will be `Coordinates.up`
  #
  # @example Set jump_force
  #   actor = Actor.new("image")
  #   actor.jump_force = 100
  attr_accessor :jump_force

  # Controls constant force this Actor will receive on each frame.
  #
  # Default `0`.
  #
  # @return [Float] the actual gravity
  #
  # @note The the direction will be `Coordinates.down`
  #
  # @example Set gravity
  #   actor = Actor.new("image")
  #   actor.gravity = 10
  attr_accessor :gravity

  # When `true` the Actor will cause and respond to collisions.
  #
  # When `false` the Actor won't cause neither respond to collisions.
  #
  # Default `true`.
  #
  # @return [Boolean] the actual solid value
  #
  # @param solid [true, false] only true or false
  #
  # @example Set solid
  #   actor = Actor.new("image")
  #   actor.solid = false
  attr_accessor :solid

  # The value to scale the image of the Actor when drawn.
  # If the value is `2` the image will rendered at double of size.
  # If the value is `0.5` the image will rendered at half of size.
  #
  # @note this value affects the attributes `width` and `height`
  #
  # Default `1`.
  #
  # @return [Float] the actual scale
  #
  # @example Set scale
  #   actor = Actor.new("image")
  #   actor.scale = 6
  attr_accessor :scale

  # The value to internal name of this Actor.
  #
  # It is useful for collision management for example.
  #
  # Default (same as `image_name`).
  #
  # @return [String] the actual name
  #
  # @example Set name
  #   actor = Actor.new("image")
  #   actor.name = "spaceship"
  #
  # @example Used in collision trigger
  #   actor = Actor.new("image")
  #   actor.on_collision do |other|
  #     if other.name == "enemy"
  #       # damage
  #     end
  #   end
  #
  # @example Used with `collision_with`
  #   actor = Actor.new("image")
  #   actor.collision_with = ["enemy", "bullet"]
  attr_accessor :name

  # In which layer the image of the Actor is rendered.
  # Smaller numbers are rendered behind higher numbers.
  #
  # Default `0`.
  #
  # @return [Integer] the actual layer
  #
  # @example Set layer
  #   actor = Actor.new("image")
  #   actor.layer = -1
  attr_accessor :layer

  # Array of strings (or "all").
  # Represents with which other solid Actors this Actor collide.
  #
  # Default `"all"`.
  #
  # @return [Array, String] the actual list of names of Actors to collide with
  #
  # @example Set with which other Actors this Actor is colliding:
  #   actor = Actor.new("image")
  #   actor.collision_with = ["enemy", "bullet"]
  #
  # @example Set this Actors collides with all other Actors
  #   actor = Actor.new("image")
  #   actor.collision_with = "all"
  attr_accessor :collision_with

  # Generate an Actor with all the default attribute values
  # @example Generate an Actor
  #   actor = Actor.new("image")
  #   actor.position # => Coordinates.zero
  #   actor.direction # => Coordinates.zero
  #   actor.speed # => 0
  #   actor.scale # => 1
  #   actor.solid # => true
  #   actor.draggable_on_debug # => true
  #   actor.layer # => 0
  #   actor.gravity # => 0
  #   actor.jump_force # => 0
  #   actor.collision_with # => "all"
  #
  # @param image_name [string] the name of the image file from `./images/*`
  # @return [Actor] the Actor
  def initialize(image_name)
    @image_name = image_name
    @image = Image.new(image_name)
    @name = image_name
    @position = Coordinates.zero
    @direction = Coordinates.zero
    @speed = 0
    @scale = 1

    @solid = true
    @draggable_on_debug = true
    @dragging = false
    @dragging_offset = nil
    @layer = 0
    @gravity = 0
    @jump_force = 0
    @collision_with = "all"

    @on_floor = false

    @on_after_move_callback = nil
    @on_collision_callback = nil
    @on_destroy_callback = nil
    @on_jumping_callback = nil
    @on_floor_callback = nil

    Global.actors << self
  end


  # Set a new image to the Actor.
  # @param image_name [String] The image file from `./images/*`
  #
  # @example Set a new image
  #   actor = Actor.new("player_walk")
  #   actor.image_name = "player_jump"
  def image=(image_name)
    @image = Image.new(image_name)
  end

  # @return [Fixnum] the Actor width in pixels
  def width
    @image.width * @scale
  end

  # @return [Fixnum] the Actor height in pixels
  def height
    @image.height * @scale
  end

  # @return [Boolean] the value of `@solid`
  def solid?
    @solid
  end

  # @param value [Coordinates] set a new direction to the Actor.
  # The vector is normalized
  #
  # @example Set a new direction
  #   actor = Actor.new("player_walking")
  #   actor.direction = Coordinates.up
  def direction=(value)
    @direction = value
    @direction = @direction.normalize unless @direction.zero?
  end

  # @!visibility private
  def draw
    @image.draw(x: position_in_camera.x, y: position_in_camera.y, scale: @scale)

    draw_debug if Global.debug
  end

  # @!visibility private
  def move
    mouse_position = Global.mouse_position + Camera.main.position

    if @draggable_on_debug && Global.debug && !@dragging && Gosu.button_down?(Gosu::MS_LEFT) && Utils.collision_at?(self, mouse_position.x, mouse_position.y)
      @dragging = true
      @dragging_offset = mouse_position - @position
    end

    if @dragging && !Gosu.button_down?(Gosu::MS_LEFT)
      @dragging = false
    end

    if @dragging
      @position = mouse_position - @dragging_offset
    else
      # Direction moving
      unless @direction.zero?
        last_position = @position
        move_by_direction

        # Check collision after cursor moving
        if @solid && @position != last_position
          manage_collisions(last_position)
        end
      end

      # Cursors moving
      last_position = @position
      move_by_cursors

      # Check collision after cursor moving
      if @solid && @position != last_position
        manage_collisions(last_position)
      end

      # Gravity force
      unless @gravity.zero?
        add_force_by_gravity
      end

      # Apply forces
      last_position = @position
      apply_forces

      # Check collision after gravity moving
      if @solid && @position != last_position
        @on_floor = false
        manage_collisions(last_position)
      end
    end

    on_after_move_do
  end

  # Destroy this Actor is not longer, moved, or rendered, or cause any collision.
  #
  # @example Destroy an Actor
  #   actor = Actor.new("image")
  #   actor.destroy
  def destroy
    on_destroy_do
    Global.actors.delete(self)
  end

  # rubocop:disable Metrics/MethodLength
  # rubocop:disable Metrics/AbcSize

  # @!visibility private
  def clone
    actor = self.class.new(@image_name)
    actor.image_name = @image_name
    actor.name = @name
    actor.position = @position.clone
    actor.direction = @direction.clone

    actor.speed = @speed
    actor.scale = @scale
    actor.solid = @solid
    actor.layer = @layer
    actor.gravity = @gravity
    actor.jump_force = @jump_force
    actor.collision_with = @collision_with

    actor.on_after_move_callback = @on_after_move_callback
    actor.on_collision_callback = @on_collision_callback
    actor.on_destroy_callback = @on_destroy_callback
    actor.on_jumping_callback = @on_jumping_callback
    actor.on_floor_callback = @on_floor_callback

    actor.on_cursor_down_callback = @on_cursor_down_callback
    actor.on_cursor_up_callback = @on_cursor_up_callback
    actor.on_cursor_left_callback = @on_cursor_left_callback
    actor.on_cursor_right_callback = @on_cursor_right_callback
    actor.on_space_bar_callback = @on_space_bar_callback
    actor.on_mouse_button_left_callback = @on_mouse_button_left_callback

    actor.on_click_callback = @on_click_callback

    actor
  end
  # rubocop:enable Metrics/AbcSize
  # rubocop:enable Metrics/MethodLength

  # Set callbacks

  # The block to be executed after each frame
  #
  # @example Limit Actor movement horizontally
  #   actor = Actor.new("image")
  #   actor.on_after_move do
  #     if actor.position.x > 100
  #       actor.position.x = 100
  #     end
  #   end
  def on_after_move(&block)
    @on_after_move_callback = block
  end

  # The block to be executed when Actor collides with another Actor
  #
  # @example Collision detected with _"bullet"_
  #   actor = Actor.new("image")
  #   actor.on_collision do |other|
  #     if other.name == "bullet"
  #       actor.destroy
  #     end
  #   end
  def on_collision(&block)
    @on_collision_callback = block
  end

  # The block to be executed before the Actor is destroyed
  #
  # @example Executes when destroyed
  #   actor = Actor.new("image")
  #   actor.on_destroy do
  #     Sound.play("explosion")
  #   end
  def on_destroy(&block)
    @on_destroy_callback = block
  end

  # The block to be executed when the Actor starts jumping
  #
  # @example Change image when jumping
  #   actor = Actor.new("walk")
  #   actor.on_jumping do
  #     actor.image_name = "jump"
  #   end
  def on_jumping(&block)
    @on_jumping_callback = block
  end

  # The block to be executed when the Actor touches floor
  #
  # @example Change image when jumping
  #   actor = Actor.new("walk")
  #   actor.on_floor do
  #     actor.image_name = "land"
  #     Clock.new { actor.image_name = "walk" }.run_on(seconds: 0.8)
  #   end
  def on_floor(&block)
    @on_floor_callback = block
  end

  # Execute callbacks

  # This method is triggered after each frame
  #
  # @example Limit Actor movement horizontally
  #   class Player < Actor
  #     def on_after_move_do
  #       if @position.x > 100
  #         @position.x = 100
  #       end
  #     end
  #   end
  def on_after_move_do
    instance_exec(&@on_after_move_callback) unless @on_after_move_callback.nil?
  end

  # This method is triggered when Actor collides with another Actor
  #
  # @example Limit Actor movement horizontally
  #   class Player < Actor
  #     def on_collision_do(other)
  #       if other.name == "bullet"
  #         destroy
  #       end
  #     end
  #   end
  def on_collision_do(other)
    instance_exec(other, &@on_collision_callback) unless @on_collision_callback.nil?
  end

  # This method is triggered before the Actor is destroyed
  #
  # @example Executes when destroyed
  #   class Player < Actor
  #     def on_destroy_do
  #       Sound.play("explosion")
  #     end
  #   end
  def on_destroy_do
    instance_exec(&@on_destroy_callback) unless @on_destroy_callback.nil?
  end

  # This method is triggered when the Actor starts jumping
  #
  # @example Change image when jumping
  #   class Player < Actor
  #     def on_jumping_do
  #       self.image_name = "jump"
  #     end
  #   end
  def on_jumping_do
    instance_exec(&@on_jumping_callback) unless @on_jumping_callback.nil?
  end

  # This method is triggered when the Actor touches floor
  #
  # @example Change image when jumping
  #   class Player < Actor
  #     def.on_floor_do
  #       self.image_name = "land"
  #       Clock.new { self.image_name = "walk" }.run_on(seconds: 0.8)
  #     end
  #   end
  def on_floor_do
    instance_exec(&@on_floor_callback) unless @on_floor_callback.nil?
  end


  protected

  attr_accessor :image_name

  # TODO: make this work optimized
  # def position_top_left
  #   @position - position_delta
  # end

  # def position_delta
  #   case @alignment
  #   when "top-left"
  #     Coordinates.zero
  #   when "center"
  #     Coordinates.new(width/2, height/2)
  #   else
  #     raise "Actor.alignment value not valid '#{@alignment}'. Valid values: 'top-left, center'"
  #   end
  # end

  def draw_debug
    Shape.rectangle(
      position: position_in_camera,
      width: width,
      height: height,
      color: Color.palette.transparent,
      stroke_color: Color.palette.red,
      stroke: 1
    ) if solid

    Global.pixel_fonts["medium"].draw_text("#{@position.x.floor},#{@position.y.floor}", position_in_camera.x, position_in_camera.y - 20, 1)
  end

  def position_in_camera
    @position - Camera.main.position
  end

  def manage_collisions(last_position)
    @velocity ||= Coordinates.zero # In case it is not initialized yet

    collisions.each do |other|
      on_collision_do(other)
      other.on_collision_do(self)

      if other.position.y >= (last_position.y + height)
        on_floor_do unless @on_floor

        @on_floor = true
        @jumping = false
        @velocity.y = 0
      end

      if other.position.y + other.height <= last_position.y
        @velocity.y = 0
      end
    end

    # # Reset position pixel by pixel
    # was_collision = false
    # while collisions.any? && @position != last_position
    #   was_collision = true
    #   last_position_direction = last_position - @position
    #   @position += last_position_direction.normalize
    # end
    # was_collision

    if collisions.any? # we don't cache collisions because position may be changed on collision callback
      @position = last_position

      return true
    end

    false
  end

  # rubocop:disable Style/Next
  def collisions
    Global.actors.reject { |e| e == self }.select(&:solid?).select do |other|
      if(
        (@collision_with == "all" || @collision_with.include?(other.name)) &&
        (other.collision_with == "all" || other.collision_with.include?(name))
      )
        Utils.collision? self, other
      end
    end
  end
  # rubocop:enable Style/Next

  protected

  attr_accessor :on_after_move_callback, :on_collision_callback, :on_destroy_callback, :on_jumping_callback, :on_floor_callback
  attr_accessor :on_click_callback
end
